sklearn/cluster/_hdbscan/_reachability.pyx

```diff
@@ -45,78 +56,121 @@ def mutual_reachability(distance_matrix, min_points=5, max_dist=0.0):
-    # should have been passed a copy.
     if issparse(distance_matrix):
-        return _sparse_mutual_reachability(
-            distance_matrix,
-            min_points=min_points,
-            max_dist=max_dist
+        # FIXME: since we convert to a CSR matrix then we do not make the operation
+        # in-place.
+        return _sparse_mutual_reachability_graph(
+            distance_matrix, min_samples=min_samples, max_distance=max_distance
         ).tocsr()
 
-    return _dense_mutual_reachability(distance_matrix, min_points=min_points)
+    return _dense_mutual_reachability_graph(distance_matrix, min_samples=min_samples)
+
 
-cdef _dense_mutual_reachability(
+cdef _dense_mutual_reachability_graph(
     cnp.ndarray[dtype=cnp.float64_t, ndim=2] distance_matrix,
-    cnp.intp_t min_points=5
+    cnp.intp_t min_samples=5
 ):
-    cdef cnp.intp_t i, j, n_samples = distance_matrix.shape[0]
-    cdef cnp.float64_t mr_dist
-    cdef cnp.float64_t[:] core_distances
+    """Dense implementation of mutual reachability graph.
+
+    The computation is done in-place, i.e. the distance matrix is modified
+    directly.
+
+    Parameters
+    ----------
+    distance_matrix : ndarray of shape (n_samples, n_samples)
+        Array of distances between samples.
+
+    min_samples : int, default=5
+        The number of points in a neighbourhood for a point to be considered
+        a core point.
+
+    Returns
+    -------
+    mututal_reachability_graph : ndarray of shape (n_samples, n_samples)
+        Weighted adjacency matrix of the mutual reachability graph. This object
+        is the same as `distance_matrix` since the operation is done in-place.
+    """
+    cdef:
+        cnp.intp_t i, j, n_samples = distance_matrix.shape[0]
+        cnp.intp_t farther_neighbor_idx = min_samples - 1
+        cnp.float64_t mutual_reachibility_distance
+        cnp.float64_t[:] core_distances
 
-    # Compute the core distances for all samples `x_p` corresponding
-    # to the distance of the k-th farthest neighbours (including
-    # `x_p`).
     core_distances = np.partition(
-        distance_matrix,
-        min_points,
-        axis=0,
-    )[min_points]
+        distance_matrix, farther_neighbor_idx, axis=0
+    )[farther_neighbor_idx]
 
     with nogil:
         for i in range(n_samples):
             for j in prange(n_samples):
-                mr_dist = max(
+                mutual_reachibility_distance = max(
                     core_distances[i],
                     core_distances[j],
-                    distance_matrix[i, j]
+                    distance_matrix[i, j],
                 )
-                distance_matrix[i, j] = mr_dist
+                distance_matrix[i, j] = mutual_reachibility_distance
     return distance_matrix
 
-# Assumes LIL format.
+
 # TODO: Rewrite for CSR.
-cdef _sparse_mutual_reachability(
+cdef _sparse_mutual_reachability_graph(
     object distance_matrix,
```

Comment:
Since this is an object, we will a huge interaction with Python anytime that we will touch `distance_matrix` in the loop below.

Basically, the way to go will be to pass the row and column indices and the data that are all NumPy arrays and we can directly access the C buffer.

Reply:
But we need a CSR or CSC format